ForumSessionProvider/ForumSessionProvider.php

```diff
@@ -18,284 +18,851 @@
+				if (preg_match('`[#<>[\]|{}@:]+`', $userName))
+					return null;
+		}
+
+		return $userName;
+	}
+
+	private function updateWikiUser()
+	{
+		$this->userWiki->setEmail($this->FSuser['email_address']);
+		$this->userWiki->mEmailAuthenticated = wfTimestampNow(); 
+		$this->userWiki->setRealName($this->FSuser['real_name']);
+
+		// Check some stuff.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+		{
+			// Convert the smf_member_id over?
+			if ($this->userWiki->getOption('forum_member_id', 0) == 0 && $this->userWiki->getOption('smf_member_id', 0) != 0)
+				$user->setOption('forum_member_id', $this->FSuser['id_member']);
+		}
+
+		$this->userWiki->setOption('forum_last_update_user', time());
+		$this->userWiki->saveSettings();
+	}
+
+	/*
+	 * Set all of our wiki groups
+	*/
+	private function setWikiGroups()
+	{
+		// Wiki Group Name => SMF Group IDS
+		$groupActions = array(
+			'sysop' => $this->FSSettings['AdminGroups'],
+//			'interface-admin' => $this->FSSettings['SuperAdminGroups'],
+			'bureaucrat' => $this->FSSettings['SuperAdminGroups'],
+		);
+
+		// Add in our special groups.
+		foreach ($this->FSSettings['SpecialGroups'] as $smf_group_id => $wiki_group_name)
+		{
+			// Group didn't exist?
+			if (!isset($groupActions[$wiki_group_name]))
+				$groupActions[$wiki_group_name] = array();
+
+			// Add the SMF group into the wiki group.
+			$groupActions[$wiki_group_name][] = $smf_group_id;
+		}
+
+		// Now we are going to check all the groups.
+		foreach ($groupActions as $wiki_group_name => $smf_group_ids)
+		{
+			// They are in the SMF group but not the wiki group?
+			if (
+				array_intersect($smf_group_ids, $this->FSgroups) != array()
+				&& in_array($wiki_group_name, $this->userWiki->getEffectiveGroups()) == array()
+			)
+				$this->userWiki->addGroup($wiki_group_name);
+			// They are not in the SMF group, but in the wiki group
+			elseif (
+				array_intersect($smf_group_ids, $this->FSgroups) == array()
+				&& in_array($wiki_group_name, $this->userWiki->getEffectiveGroups()) != array()
+			)
+				$this->userWiki->removeGroup($wiki_group_name);
+		}
+
+		// Did we make any changes?
+		$this->userWiki->setOption('forum_last_update_groups', time());
+		$this->userWiki->saveSettings();
+	}
+
+	/*
+	 * Create a Wiki user
+	*/
+	private function createWikiUser()
+	{
+		$this->userWiki->setName($this->FSuserName);
+		$this->userWiki->setEmail($this->FSuser['email_address']);
+		$this->userWiki->setRealName($this->FSuser['real_name']);
+		$this->userWiki->mEmailAuthenticated = wfTimestampNow();
+
+		$this->userWiki->addToDatabase();
+
+		// Only SMF does this.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+			$user->setOption('forum_member_id', $this->FSuser['id_member']);
+
+		$this->userWiki->setOption('forum_last_update', time());
+		$this->userWiki->saveSettings();
+	}
+
+	// !!! Add checks here for banning
+	private function isBanned()
+	{
+		switch ($this->wgFSPSoftware)

```

Comment:
I intentionally left this out. Elkarte's banning schema is largely the same as SMF's is for now.

I'd prefer an option to check for bans. The switch can remain but can safely include everyone for now.

Reply:
Sounds reasonable.  I don't know elks db to know what is compatible.

Reply:
The only difference is how IPs are handled, slightly.

I'm planning on a more comprehensive IP replacement which SMF may want to adopt but that's a whole 'nother discussion.